### DISCUSSION TITLE
Call function in kernel or materials

### DISCUSSION THREAD

**Discussion by kunokchang** [idaholab/moose#16297] [Q&A Modules: Phase field] [answered]
I posted a question yesterday, but it seems to be lacking in explanation, so I will replenish it.

I would like to call our custom function in kernel or materials.

Materials]
  [./bulk_free_energy]
    type = DerivativeParsedMaterial
    f_name = fb
    args = 'eta1'
    constant_names = 'a'
    constant_expressions = '2.057'
    function = 'a*((eta1)^2'
    derivative_order = 2
  [../]

I need to add result of Fermi-Dirac function like a*((eta1)^2+F_{1/2}(eta1)

Which is given here

[Functions]
  [./fermi]
    type = PiecewiseLinear
    data_file = fermi.csv
    format = columns
  [../]
[]

fermi.csv is given by (part of it)

  -7.9899997711181641        2.1235612216047891E-004
  -7.9800000190734863        2.1449011174362248E-004
  -7.9699997901916504        2.1664564757151034E-004
  -7.9600000381469727        2.1882273951932466E-004
  -7.9499998092651367        2.2102181283188398E-004
  -7.9400000572204590        2.2324287758167501E-004
  -7.9299998283386230        2.2548636760039243E-004
  -7.9200000762939453        2.2775229316172307E-004
  -7.9099998474121094        2.3004109685755749E-004
  -7.9000000953674316        2.3235278916675431E-004
  -7.8899998664855957        2.3468782161821161E-004

But we are not clear how can we call result of Fermi-Dirac function with argument eta1. 

If there is a better way to do this, we would really appreciate any comments.

**Comment by laagesen** [comment, marked as answer]
Hi Kunok, I just commented on the other post, but let me amplify a little bit based on the extra detail in this post. Since you want the Fermi-Dirac distribution as a function of eta1, you can just add that term to the function you are defining in your DerivativeParsedMaterial, using the analytical form shown here:

https://www.doitpoms.ac.uk/tlplib/semiconductors/fermi.php

**Reply by kunokchang** [reply]
Dear Laagesen 

Sorry for confusion we cause. We want to calculate Fermi-Dirac integration, not distribution function.

The analytic form is given below and take integration is not a trivial job so we try to interpolate it. 

https://en.wikipedia.org/wiki/Complete_Fermi%E2%80%93Dirac_integral

**Reply by kunokchang** [reply]
We try to evaluate F_{1/2}(eta)

**Reply by laagesen** [reply]
I see-  so are you assuming F_{1/2}(eta) is not changing at each time step and using the same .csv file at each time step? Or are you wanting to output eta, interpolate it using the method that is generating the .csv file, then read it back in?

**Reply by kunokchang** [reply]
We assume F_{1/2}(eta) is not changing at each time step and using the same .csv file, however, eta is time-dependent term.

**Reply by laagesen** [reply]
Hm, I don't think there is a way to do what you are wanting to with the `PiecewiseLinear` function. `PiecewiseLinear` assumes the first part of the data in the csv file is either time (by default) or x,y, or z positions (with optional argument). See https://mooseframework.inl.gov/source/functions/PiecewiseLinear.html for more details on this. It's not set up to take problem variables as an argument.

An alternative that I think would work would be to use the `PiecewiseLinearInterpolationMaterial`. See here for details and examples. https://mooseframework.inl.gov/source/materials/PiecewiseLinearInterpolationMaterial.html 
You would use this to create a material property that depends on `eta1`. Unfortunately it does not look like it has the capability to read in a csv file to get the x-y pairs, so you would have to specify them in the input file (although you could probably add the functionality to read csv files pretty easily using the code from the `PiecewiseLinear` function as an example). So `PiecewiseLinearInterpolationMaterial` would create a material property, say it's called F. You would then make that available to your existing `DerivativeParsedMaterial` called `bulk_free_energy` using `material_property_names = F`. 

It does look like `PiecewiseLinearInterpolationMaterial` calculates derivatives, which I'm assuming you're going to need since most phase-field models need the derivative of the free energy. However just be aware that since you are using linear interpolation, all of your second derivatives are going to be zero; these are used to calculate Jacobians by the preconditioners, so convergence in your solve may be hurt.

**Reply by kunokchang** [reply]
Sorry for delayed response. I gave my grades at the end of the semester, and now I am back with this problem.

The answers you gave are very helpful in approaching the problem. I would like to express our sincere thanks.

We declare PiecewiseLinearInterpolationMaterial property as below and we want to call it in DerivativeParsedMaterial and is it possible to pass fermi_test(eta1/2.0) like this way? We are not clear how to pass argument of function. 

[./fermi_test]
    type = PiecewiseLinearInterpolationMaterial
    property = fermi_test
    variable = fermi
    x = '2.3235278916675431E-004 2.5678716657140193E-004 2.8379094144941774E-004 3.1363395344228466E-004 3.4661480184293897E-004 3.8306332338290775E-004 4.2334400591023457E-004 4.6785984730409537E-004 5.1705550309345436E-004 5.7142299341010746E-004 6.3150577678226284E-004 6.9790436428065117E-004 7.7128266457680090E-004 8.5237315288779500E-004 9.4198625416296500E-004 1.0410170287445100E-003 1.1504543904029938E-003 1.2713915208859713E-003 1.4050343362219744E-003 1.5527168865464539E-003 1.7159122952076673E-003 1.8962478534793204E-003 2.0955220648618544E-003 2.3157184525443913E-003 2.5590307589031741E-003 2.8278807548148058E-003 3.1249428476409127E-003 3.4531718412657711E-003 3.8158252930559911E-003 4.2165045470864424E-003 4.6591834873861418E-003 5.1482483588226420E-003 5.6885426181912955E-003 6.2854027000215241E-003 6.9447242909850462E-003 7.6730081043849289E-003 8.4774235261575424E-003 9.3658801128583824E-003 1.0347083644346118E-002 1.1430641727716809E-002 1.2627134940562407E-002 1.3948216455171785E-002 1.5406723459013859E-002 1.7016761909108134E-002 1.8793871039066580E-002 2.0755129368957407E-002 2.2919305158910545E-002 2.5307023402706061E-002 2.7940885872608706E-002 3.0845717761828609E-002 3.4048713809169372E-002 3.7579651543448035E-002 4.1471124785238581E-002 4.5758692787145720E-002 5.0481226394378478E-002 5.5681079412960861E-002  6.1404356606654949E-002  6.7701199554686134E-002  7.4625918988119933E-002  8.2237426265194619E-002  9.0599359529987766E-002  9.9780341016472748E-002 0.10985423431455218 0.12090012503482848 0.13300274604701343 0.14625237774326805 0.16074492035514423 0.17658192596875985 0.19387015748734349 0.21272181018402139 0.23325387611972009 0.25558775599966849 0.27984879160089887 0.30616507367908830 0.33466728239940641 0.36548725419244205 0.39875699587613084 0.43460766605346784 0.47316760840632061 0.51456206161858054 0.55891138443652488 0.60632883599122311 0.65692179481661872 0.71078722503159686 0.76801459136108396 0.82868157559417333 0.89285491195281386 0.96059203187739461 1.0319365775923359 1.1069236669593814 1.1855755056991080 1.2679037300406737 1.3539125176629796 1.4435937108406798 1.5369341623179518 1.6339106689096636 1.7344932376333677 1.8386489768331284 1.9463361365213048 2.0575128598853629 2.1721308396961607 2.2901389570235700 2.4114874479487196 2.5361205226177570 2.6639861303092491 2.7950281621238084 2.9291902642563219 3.0664201414726628 3.2066608072888370 3.3498612866938982 3.4959675214463499 3.6449264163040018 3.7966903498567839 3.9512072564175238 4.1084322668677187 4.2683174995857973 4.4308163880114391 4.5958884506057514 4.7634883782231263 4.9335786157998385 5.1061182183151681 5.2810675193328223 5.4583932138892495 5.6380565883668146 5.8200270067518005 6.0042699264614274 6.1907518921002884 6.3794459195002933 6.5703189478053332 6.7633461920765487 6.9584984147593936 7.1557472593657199 7.3550709491787369 7.5564410324326055 7.7598375481000117 7.9652355894842275 8.1726109369660396 8.3819460365873244 8.5932160735643013 8.8064049118334271 9.0214910110319586 9.2384533674000124 9.4572777810297612 9.6779423044161685 9.9004339103835033 10.124733776288622 10.350823490228317 10.578691569533502 10.808318301835646 11.039693111246226 11.272799257988478 11.507620323701010 11.744146994171068 11.982361330444245 12.222254797793555 12.463812384186419 12.707019323593313 12.951868106324753 13.198342188637293 13.446434662421453 '
    y = '-7.9 -7.8 -7.7 -7.6 -7.5 -7.4 -7.3 -7.2 -7.1 -7.0 -6.9 -6.8 -6.7 -6.6 -6.5 -6.4 -6.3 -6.2 -6.1 -6.0 -5.9 -5.8 -5.7 -5.6 -5.5 -5.4 -5.3 -5.2 -5.1 -5.0 -4.9 -4.8 -4.7 -4.6 -4.5 -4.4 -4.3 -4.2 -4.1 -4.0 -3.9 -3.8 -3.7 -3.6 -3.5 -3.4 -3.3 -3.2 -3.1 -3.0 -2.9 -2.8 -2.7 -2.6 -2.5 -2.4 -2.3 -2.2 -2.1 -2.0 -1.9 -1.8 -1.7 -1.6 -1.5 -1.4 -1.3 -1.2 -1.1 -1.0 -0.9 -0.8 -0.7 -0.6 -0.5 -0.4 -0.3 -0.2 -0.1 0.0 0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.0 1.1 1.2 1.3 1.4 1.5 1.6 1.7 1.8 1.9 2.0 2.1 2.2 2.3 2.4 2.5 2.6 2.7 2.8 2.9 3.0 3.1 3.2 3.3 3.4 3.5 3.6 3.7 3.8 3.9 4.0 4.1 4.2 4.3 4.4 4.5 4.6 4.7 4.8 4.9 5.0 5.1 5.2 5.3 5.4 5.5 5.6 5.7 5.8 5.9 6.0 6.1 6.2 6.3 6.4 6.5 6.6 6.7 6.8 6.9 7.0 7.1 7.2 7.3 7.4 7.5 7.6 7.7 7.8 7.9 8.0'
  [../]

  [./bulk_free_energy]
    type = DerivativeParsedMaterial
    f_name = fb
    args = 'eta1 eta2 eta3 eta4 mu1 mu2 mu3 mu4'
    constant_names = 'a aa temp temp_c T0_eta T0_mu b11 b13 c11 c13 bb11 bb13 cc11 cc13 h'
    constant_expressions = '2.057 3.943 280 338 275 270 -0.623 0.121 0.331 4.189 1.368 -3.679 0.400 2.000 0.300'
    material_property_names = 'fermi_test'
    function = 'a*(temp-T0_eta)/(2.*temp_c)*((eta1)^2+fermi_test'
    derivative_order = 2
  [../]

**Reply by laagesen** [reply]
From what you said earlier, you are wanting to interpolate F_{1/2}(eta)  - so you would need to specify `variable = eta` in the `[./fermi_test]` block (or whichever order parameter you want to interpolate as a function of, since it looks like there are multiple order parameters). Right now you are specifying `variable = fermi` and I'm not sure what that is. If the derivatives of `fb` are going to be used in one of your kernels, you will also want to specify `material_property_names = fermi_test(eta1)` in `[./bulk_free_energy]` (assuming `eta1` is what you are interpolating based on) rather than just `material_property_names = fermi_test`- sorry I probably did not mention that before. This will make sure that the derivatives of `fermi_test` are accounted for when derivatives of `fb` are taken, as detailed here: https://mooseframework.inl.gov/source/materials/DerivativeParsedMaterial.html . It may also be helpful to output the property `fermi_test` to your exodus file by specifying `outputs = exodus` in the `[./fermi_test]` block so you can visualize it and make sure you are getting what you expect.

**Reply by kunokchang** [reply]
We really appreciate for your help. It has been a huge help in applying the MOOSE framework to our research.

**Reply by laagesen** [reply]
Great glad we could help!
